sdv/metadata/single_table.py

```diff
@@ -538,20 +538,72 @@ def _determine_sdtype_for_objects(self, data):
 
         return sdtype
 
+    def _detect_primary_key(self, data):
+        """Detect the table's primary key.
+
+        This method will loop through the columns and select the first column that was detected as
+        an id. If there are none of those, it will pick the first unique pii column. If there is
+        still no primary key, it will return None. All other id columns after the first will be
+        reassigned to the 'unknown' sdtype.
+
+        Args:
+            data (pandas.DataFrame):
+                The data to be analyzed.
+
+        Returns:
+            str:
+                The column name of the selected primary key.
+
+        Raises:
+            RuntimeError:
+                If the sdtypes for all columns haven't been detected or set yet.
+        """
+        original_columns = data.columns
+        stringified_columns = data.columns.astype(str)
+        data.columns = stringified_columns
```

Comment:
Do we need to do this here and in `_detect_columns`?

Reply:
I think so because the column names that get added to `self.columns` seem to be the [stringified ones](https://github.com/sdv-dev/SDV/blob/2abacb192192182cbe9a712f8e5df3d75db3c480/sdv/metadata/single_table.py#L592). So if we don't do it here then we might get a `KeyError`

Reply:
Actually now that I think about it, this could be a problem for every other function that takes data and looks in the columns. Maybe we should change the dict to store the original column names instead

Reply:
But we modified the data so the column names are the stringified ones, no? Then at the end we convert it back to the original column names.

Reply:
We also do it at the beginning of `_detect_columns` and then restore them at the end

Reply:
Yeah but the idea is eventually people will call this method on their own